design/observerability.md

```diff
@@ -1,150 +1,263 @@
+This way or another, the implementation should use either Micrometer directly or [Smallrye Metrics](https://github.com/smallrye/smallrye-metrics).
+The former approach is much more preferable.
 
-Exactly what additional metrics we should expose in the first round is still to be decided.
+At the time of writing this design, Quarkus already provides [Micrometer extension](https://quarkus.io/guides/micrometer).
 
-We will start small with the aim to expand on the available metrics in the future.
+Depending on timing, it is very likely that Wildfly deployment will require embedding Micrometer jars into Keycloak's modules
+and configuring them manually. Deploying Micrometer as Keycloak's dependency also has some impact on productization. The current
+version of Micrometer brings in the following dependencies:
+* `org.hdrhistogram:HdrHistogram`
+* `org.latencyutils:LatencyUtils`
+* `io.prometheus:simpleclient_common`, which also brings `io.prometheus:simpleclient_pushgateway`
 
-### Logging
+Depending on the timing, the Wildfly's distribution may need to expose all metrics in its own application context:
 
-WildFly covers JSON formatted logging. We will simply add some documentation.
+    http://127.0.0.1:8080/auth/metrics
 
-### Tracing
+In this case, it will be necessary to develop a custom bridge between JCA Metrics and the endpoint exposed in Keycloak. The
+[Healthcheck API Design](https://issues.redhat.com/browse/KEYCLOAK-12398) requires some of them to inspect if the database
+is up and running. The implementation of such a bridge might be similar to the
+[Hibernates's Metrics implementation](https://github.com/micrometer-metrics/micrometer/blob/0dc111d1f6d3aa288dda24ac0912cbb483f1da1e/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jpa/HibernateMetrics.java#L191).
 
-Most transactions to Keycloak are directly from the browser and do not invoke other services. Applications invoke Keycloak
-to exchange code for a token, to refresh tokens and finally services may invoke Keycloak to verify tokens. These are
-simpler use-cases as we haven't seen much demand for tracing in Keycloak. Hence distributed tracing is a low priority
-and will be considered in the future. 
+Both implementations (Wildfly and Quarkus based) will use a similar process for implementing metrics. The Micrometer
+has a notion of a [Gloabl Registry](https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/Metrics.java#L35)
+that can be used for environments where CDI injection is not possible (as a reminder, Keycloak doesn't use CDI). The implementation
+can obtain all metrics instruments (such as gauges, histograms etc.) from it.
 
+## Metrics for the REST interface
 
-## Resources
+Most of the metrics are related to the various REST endpoints. Therefore, it makes sense to put the root metrics
```

Comment:
Depending on what library you use for REST endpoints, there might already be a Micrometer Binder that's applicable to automatically provide those metrics.

If not, it might be possible to utilize what has been done for Quarkus, if you're using Vert.x underneath.

As a last resort, developing a `MeterBinder` directly shouldn't be too hard